### PR TITLE
updates to image acquisition

### LINKED ISSUE #129: Changes to Imaging Acquisition schema
Some suggested changes and questions to the acquisition schema (up for discussion) are as follows:

CHANGES
* Add _unit to end of any field name that has tangible units associated with it.
* Remove voxel_size from Axis() and fold it into Tile().
* Remove step_size_z from Acquisition() and fold it into Tile().
* Change tile_overlap from percent to microns.
* Remove enabled field from Channel (not needed). Implicitly enabled if it is listed as a channel.
* Remove the tile_prefix since I think this won't be needed (see below).
* We restructure positions and channels within a tiles class. 
* Nest Position, Channel, DAQ, Notes --> within Tile.
     * Tiles
└── Filename
└── Voxel size --> this contains voxel_size_x_um, voxel_size_y_um, voxel_size_z_um
└── Position
└── Channel
└── DAQ
└── Notes
     * Add a field for filename string which will point to the datafile name corresponding to this tile.
     * Rename positions to position, and make change to Position instead of List[Position]
     * Rename channels to channel, and make change to Channel instead of List[Channel]
     * Rename daqs to daq, and make change to dict instead of List[dict]
     
QUESTIONS
* Do we coordinate the Axis name with the instrument schema? In the instrument schema we have stage_axis_direction which is instrument specific. Also stage_axis_name which maps to the lettered axis of the system's stage. Within our acquisition software we remap stage_axis_name to what is essentially acquisition_axis_names. Acquisition_axis_names maps to stage_axis_direction but for readability and verbose reasons, we do not use the terms "detection axis", "illumination axis", and "perpendicular axis" but instead "Z", "X", and "Y" respectively. So the mapping goes something like X --> Z, Y --> Z, Z --> Y. How should we handle this in the acquisition schema Axis class?
* Consider removing additional parameters, seems overlapping with notes?
* Consider removing tile overlap from Acquisition(). It is captured more indirectly in the position of each tile. Unless it is handy to be able to quickly know how much overlap was used in the dataset?
* Consider removing volume from Axis(). Same as above, unless at quick glance it is useful to know how large the overall dataset volume is?
Example envisioned code below...
```python
""" example ExaSPIM acquisition """
from aind_data_schema.imaging import acquisition

acq=acquisition.Acquisition(
    experimenter_full_name="###", # TODO. need to add this to the GUI/config
    subject_id="###", # TODO. need to grab this from the config
    instrument_id="###", # TODO. is this needed?
    session_start_time="###", # TODO. fill this field at start of record function
    session_end_time="###", # TODO.fill this field at end of record function
    local_storage_directory="D:", # TODO. grab this from config 
    external_storage_directory="Z:", # TODO. grab this from config 
    axes=[
        acquisition.Axis(
            name="X", # Tiger motor axis. Do we coordinate this with the conventions in the instrument schema?
            dimension=2, # Imaris file axis dimension -> informed by sample pose remapping
            direction="Left_to_right", # TODO. Think about how we capture this.
        ),
        acquisition.Axis(
            name="Y", # Tiger motor axis. Do we coordinate this with the conventions in the instrument schema?
            dimension=1, # Imaris file axis dimension -> informed by sample pose remapping
            direction="Anterior_to_posterior", # TODO. Think about how we capture this.
        ),
        acquisition.Axis(
            name="Z", # Tiger motor axis. Do we coordinate this with the conventions in the instrument schema?
            dimension=0, # Imaris file axis dimension -> informed by sample pose remapping
            direction="Inferior_to_superior", # TODO. Think about how we capture this.
        )
    ]
    tiles=[
        acquisition.Tile(
            filename="tile_X_0000_Y_0000_Z_0000_CH_488.ims", # TODO. grab this from config.
            voxel_size=[
                voxel_size_x_um=0.748, # TODO. grab this from config.
                voxel_size_y_um=0.748 # TODO. grab this from config.
                voxel_size_z_um=1.000 # TODO. grab this from config.
            ],
            position=acquisition.Position(
                x_start_um=0, # TODO. grab this from config.
                x_end_um=100, # TODO. grab this from config.
                y_start_um=0, # TODO. grab this from config.
                y_end_um=100, # TODO. grab this from config.
                z_start_um=0, # TODO. grab this from config.
                z_end_um=100 # TODO. grab this from config.
            ),
            channel=acquisition.Channel(
                channel=488, # TODO. grab this from config.
                laser_power=200, # TODO. grab this from config.
                filter_wheel_index=0 # TODO. grab this from config.
            ),
            daq=[
                etl_buffer_time_ms=5.0, # TODO. grab this from config.
                etl_amplitude_volt=0.2, # TODO. grab this from config.
                etl_offest_volt=1.77 # TODO. grab this from config.
            ],
            notes="these are my notes"
        ),
        acquisition.Tile(
            filename="tile_X_0000_Y_0000_Z_0000_CH_561.ims",
            voxel_size=[
                voxel_size_x_um=0.748, # TODO. grab this from config.
                voxel_size_y_um=0.748 # TODO. grab this from config.
                voxel_size_z_um=1.000 # TODO. grab this from config.
            ],
            position=acquisition.Position(
                x_start_um=0, # TODO. grab this from config.
                x_end_um=100, # TODO. grab this from config.
                y_start_um=0, # TODO. grab this from config.
                y_end_um=100, # TODO. grab this from config.
                z_start_um=0, # TODO. grab this from config.
                z_end_um=100 # TODO. grab this from config.
            ),
            channel=acquisition.Channel(
                channel=561, # TODO. grab this from config.
                laser_power=200, # TODO. grab this from config.
                filter_wheel_index=0 # TODO. grab this from config.
            ),
            daq=[
                etl_buffer_time_ms=5.0, # TODO. grab this from config.
                etl_amplitude_volt=0.2, # TODO. grab this from config.
                etl_offest_volt=1.77 # TODO. grab this from config.
            ],
            notes="these are my notes"
        ),
    ]
)

inst.write_standard_file(prefix="exaspim_acquisition")
```

### PR DESCRIPTION
**Pull request by dyf**
resolves #129

### CODE REVIEW

**Comment by dyf** [comment]
@adamkglaser @saskiad I tried a different way of handling units so they show up in the JSON. Free to say no.

**Comment by adamkglaser** [comment]
These updates look good to me. 
* I think it is good for human readability to dump units into the JSON. 
* Where is this being handled -> in the base class? Laser power should also have a unit associated with it -> mW.
* For the "DAQ" parameters since these will differ (or be non-existent) for some other imaging datasets, we can just have _unit at the end of the variable name for the units? These will not need to be ingested by any post-processing codes. Tracking these is more for completeness and for debugging purposes if needed.
* Should we stick with having the axis letter first in variable names? So size_x -? x_size etc.?
* Is it needed to specify how the axes are physically being used? For all SPIM systems I think this would break down into "slow tiling axis", "fast tiling axis", "scanning axis".
* I agree with your comment that lets just stick with X, Y, Z in the acquisition mapping to X, Y, Z in the instrument schema. And let's just ignore our remapping of hardware stage lettered axes to the X, Y, Z in the instrument schema. It can maybe be mentioned in the notes of the motorized stages of the instrument schema.
* I agree with @miketaormina that immersion medium should be captured in the acquisition schema. Perhaps as an added field in Acquisition(). This could be a ```str``` for the medium name, and perhaps also a ```float``` for the measured RI of the medium.
* For the smartSPIM as well, since there are potentially multiple objective lenses, we should also add a field for the objective lens used for this particular acquisition. What is the best way to handle this to map consistently with the instrument schema?
* Thinking forward to the diSPIM, we should add a field for the angle between the light-sheet and the "scanning axis" defined above. For the exaSPIM/smartSPIM this is 90 deg. but is 45 deg for the diSPIM. Presumably we'd want to capture this in such a way that it could be pulled by post-processing code to know how to apply deskewing. Based on this I think we should add a field to Tile() for light_sheet_angle (?) in units of deg.

**Comment by adamkglaser** [comment]
Looking ahead towards integrating with OME-Zarr. We should:

1. Make sure that our ```Axis``` definition has enough information to unambiguously enable definition of axes as specified by OME-NGFF. https://ngff.openmicroscopy.org/latest/#axes-md.
2. Make sure that our ```TilePosition``` definition similar can map onto the ```Translation``` coordinate transform specified by OME-NGFF. https://ngff.openmicroscopy.org/latest/#trafo-md.
3. Make sure that our ```VoxelSize``` definition can map onto the ```Scale``` coordinate transform specified by OME-NGFF. https://ngff.openmicroscopy.org/latest/#trafo-md.
4. We can then store the rest of our acquisition JSON file with a ```layout``` we decide upon, similar to the scheme proposed by bioformats. https://ngff.openmicroscopy.org/latest/#bf2raw.

**Review by miketaormina — APPROVED** [review]
This all looks reasonable to me.

Edit: I would have restricted approval if I had known only one of us was required. ExaSPIM edits should be approved by Adam.

**Review by adamkglaser — APPROVED** [review]
Looks good to me as well!

File: examples/exaspim_acquisition.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    tiles=[
+        acquisition.Tile(
+            file_name="tile_X_0000_Y_0000_Z_0000_CH_488.ims",
+            coordinate_transformations=[
+                acquisition.Scale3dTransform(scale=[0.748, 0.748, 1.0]),
+                acquisition.Translation3dTransform(translation=[0, 0, 0]),
+            ],
+            channel=acquisition.Channel(
+                channel_name="488",
+                laser_wavelength=488,
```
Comment (saskiad) [inline examples/exaspim_acquisition.py:44]: this needs units. same with laser_power below

Comment (saskiad) [reply examples/exaspim_acquisition.py:44]: oh - or not? this is the .py file.

File: src/aind_data_schema/imaging/acquisition.py [hunk 2]
```diff
@@ -82,53 +72,68 @@ def from_direction_code(code, voxel_sizes, volume_sizes) -> List[Axis]:
...
-    tile_prefix: Optional[str] = Field(
+    tiles: List[Tile] = Field(..., title="Acquisition tiles")
+    axes: List[Axis] = Field(..., title="Acquisition axes")
+    chamber_immersion: Immersion = Field(..., title="Acquisition chamber immersion data")
+    sample_immersion: Optional[Immersion] = Field(None, title="Acquisition sample immersion data")
+    active_objectives: Optional[List[str]] = Field(
         None,
-        description="Zstacks will be named: <tile_prefix>_<x>_<y>_<wavelength>.tiff",
-        title="Tile prefix",
+        title="List of objectives used in this acquisition. If null, all objectives in instrument were used.",
```
Comment (saskiad) [inline src/aind_data_schema/imaging/acquisition.py:136]: I really am opposed to secret codes. If all objectives are used, there should be a list of all objectives.

File: src/aind_data_schema/imaging/acquisition.py [hunk 3]
```diff
@@ -82,53 +72,68 @@ def from_direction_code(code, voxel_sizes, volume_sizes) -> List[Axis]:
...
+    file_name: Optional[str] = Field(None, title="File name")
+    notes: Optional[str] = Field(None, title="Notes")
+    imaging_angle: int = Field(0, title="Imaging angle")
+    imaging_angle_unit: str = Field("degree", title="Imaging angle unit")
+
+
+class Immersion(AindModel):
+    """Description of immersion media"""
 
-    x_start_um: float
```
Comment (saskiad) [inline src/aind_data_schema/imaging/acquisition.py:95]: did this info move somewhere else that I don't see?

File: src/aind_data_schema/imaging/acquisition.py [hunk 4]
```diff
@@ -82,53 +72,68 @@ def from_direction_code(code, voxel_sizes, volume_sizes) -> List[Axis]:
...
+    axes: List[Axis] = Field(..., title="Acquisition axes")
+    chamber_immersion: Immersion = Field(..., title="Acquisition chamber immersion data")
+    sample_immersion: Optional[Immersion] = Field(None, title="Acquisition sample immersion data")
+    active_objectives: Optional[List[str]] = Field(
         None,
-        description="Zstacks will be named: <tile_prefix>_<x>_<y>_<wavelength>.tiff",
-        title="Tile prefix",
+        title="List of objectives used in this acquisition. If null, all objectives in instrument were used.",
     )
-    tile_overlap_x: Optional[float] = Field(None, title="Tile overlap x (percent)", ge=0, le=100)
```
Comment (saskiad) [inline src/aind_data_schema/imaging/acquisition.py:126]: ditto this section? are we just not using it anymore? Don't we need it?

File: src/aind_data_schema/imaging/acquisition.py [hunk 5]
```diff
@@ -82,53 +72,68 @@ def from_direction_code(code, voxel_sizes, volume_sizes) -> List[Axis]:
...
-    """Description of stage position"""
+class Scale3dTransform(AindModel):
+    """3D scale transform"""
+
+    type: str = Field("scale", title="transformation type")
+    scale: conlist(float, min_items=3, max_items=3) = Field(..., title="3D scale parameters")
+
+
+class Translation3dTransform(AindModel):
+    """3D scale transform"""
```
Comment (saskiad) [inline src/aind_data_schema/imaging/acquisition.py:91]: these classes need better descriptions to explain what they are

File: src/aind_data_schema/imaging/acquisition.py [hunk 6]
```diff
@@ -82,53 +72,68 @@ def from_direction_code(code, voxel_sizes, volume_sizes) -> List[Axis]:
...
     filter_wheel_index: int = Field(..., title="Filter wheel index")
 
 
-class Position(AindModel):
-    """Description of stage position"""
+class Scale3dTransform(AindModel):
+    """3D scale transform"""
+
+    type: str = Field("scale", title="transformation type")
+    scale: conlist(float, min_items=3, max_items=3) = Field(..., title="3D scale parameters")
```
Comment (saskiad) [inline src/aind_data_schema/imaging/acquisition.py:87]: add units

**Review by saskiad — CHANGES_REQUESTED** [review]
minor changes

File: src/aind_data_schema/imaging/acquisition.py (hunk 6 above)
Comment (dyf) [reply src/aind_data_schema/imaging/acquisition.py:87]: scale is unitless. The `Axis` model has units.

File: src/aind_data_schema/imaging/acquisition.py (hunk 2 above)
Comment (dyf) [reply src/aind_data_schema/imaging/acquisition.py]: deal

File: src/aind_data_schema/imaging/acquisition.py (hunk 5 above)
Comment (dyf) [reply src/aind_data_schema/imaging/acquisition.py]: done

File: src/aind_data_schema/imaging/acquisition.py (hunk 4 above)
Comment (dyf) [reply src/aind_data_schema/imaging/acquisition.py:126]: this is handled by coordinate transformations now

**Review by saskiad — APPROVED** [review]

**Pull request merged**